python/packages/nisar/products/writers/BaseL2WriterSingleInput.py

```diff
@@ -42,6 +42,52 @@
 LUT_1D_RG_DATASETS_LUT_EXPANSION_ALONG_AZ_MARGIN_IN_PIXELS = 5
 
 
+def get_static_layers_data_access(static_layers_data_access_runconfig,
+                                  granule_id):
+    """
+    Read static layers data access template from the input runconfig field
+    `static_layers_data_access` and replace placeholder(s) with
+    the actual values.
+
+    Parameters
+    ----------
+    static_layers_data_access_runconfig: scalar
+        Value from runconfig field
+        `ceos_analysis_ready_data.static_layers_data_access`
+    granule_id: str
+        Product's granule ID
+
+    Returns
+    -------
+    static_layers_data_access: str
+        An URL representing the static layers data access with all
+        placeholders replaced. Returns "(NOT SPECIFIED)" if
+        `static_layers_data_access_runconfig` is None or an empty
+        string.
+    """
+
+    if not static_layers_data_access_runconfig:
+        return '(NOT SPECIFIED)'
```

Comment:
Let's simplify the logic of this function, and establish very clear cases and delineate the behavior for each case. Something like:

```suggestion
    default = "(NOT SPECIFIED)"
    if (template_string is None):
        log.error(...)
        return default
    elif (granule_id is None) or (granule_id == "(NOT SPECIFIED)"):
        log.error(...)
        return default
    elif "{granule_id}" not in template_string:
        log.error(..)
        return default
    else:
        return template_string.replace('{granule_id}', granule_id)
```

Reply:
This logic doesn’t look correct. Logging the error may cause pyre to raise an exception, which conflicts with the `return default`.

Reply:
Oof, I didn't know that about pyre. Fair enough...

Could we make these warnings then?

Reply:
More importantly, there are edge cases missing from the updated code.

Here's one example: What happens if the template string "https://my.static.link/granule_id.h5" is passed in, and also the user provides an actual granule ID? If I'm the user, I would expect the function to make use of the granule ID. Instead, it silently ignores it.

Later on, someone will look at the HDF5 file and wonder why the Dataset is populated with "https://my.static.link/granule_id.h5". There is no indication of the SAS' behavior in either the docstring nor the log output, so this will be very hard to debug.

Reply:
Thank you, @nemo794 , I don't think we need to force the user to include the substring `{granule_id}. Notice that they can provide the final URL or even a DOI. 

If they make a mistake and forget the curly braces, I don't think it should be the responsibility of the SAS to try to figure this out. Notice that the "replace" behavior  is the same strategy that we used for populating the partial granule ID (example [here](https://github.com/isce-framework/isce3/blob/711ccd8798c61b229b97022d726ddf4305fa72e3/python/packages/nisar/products/writers/BaseWriterSingleInput.py#L132), [here](https://github.com/isce-framework/isce3/blob/711ccd8798c61b229b97022d726ddf4305fa72e3/python/packages/nisar/products/writers/SLC.py#L346), and [here](https://github.com/isce-framework/isce3/blob/711ccd8798c61b229b97022d726ddf4305fa72e3/python/packages/nisar/products/insar/granule_id.py#L239)), and also OPERA RTC-S1 ([here](https://github.com/opera-adt/RTC/blob/c5cc7403af16bb67fd14f43bf2aec15c6a422297/src/rtc/h5_prep.py#L485) and [here](https://github.com/opera-adt/RTC/blob/c5cc7403af16bb67fd14f43bf2aec15c6a422297/src/rtc/rtc_s1_single_job.py#L59))

Reply:
Let me revise the example: what happens if I pass in the template string "https://my.static.link/static.h5", and I also pass in an actual granule ID? As the user, I would expect the function to make use of the granule ID I'm intentionally providing as an input parameter. Instead, the function as-is ignores the granule ID silently.

Reply:
Your example is fine. There's no need to substitute anything there. In the output product you'd have: `staticLayersDataAccess = b"https://my.static.link/static.h5"`